client/my-sites/site-settings/jetpack-credentials/credentials-form/index.jsx

```diff
@@ -0,0 +1,280 @@
+import FormTextArea from 'components/forms/form-textarea';
+import FormInputValidation from 'components/forms/form-input-validation';
+
+export class CredentialsForm extends Component {
+	state = {
+		showPublicKeyField: false,
+		form: {
+			protocol: 'ssh',
+			host: '',
+			port: '',
+			user: '',
+			pass: '',
+			abspath: '',
+			kpub: '',
+		},
+		formErrors: {
+			host: false,
+			port: false,
+			user: false,
+			pass: false,
+		},
+	};
+
+	componentWillReceiveProps( nextProps ) {
+		const changedFields = [ 'protocol', 'host', 'port', 'user' ].filter(
+			key => nextProps[ key ] !== this.props[ key ]
+		);
+
+		const changedForm = changedFields.reduce(
+			( props, key ) => ( { ...props, [ key ]: nextProps[ key ] } ),
+			{}
+		);
+		const resetErrors = changedFields.reduce(
+			( props, key ) => ( { ...props, [ key ]: false } ),
+			{}
+		);
+
+		this.setState( {
+			form: { ...this.state.form, ...changedForm },
+			formError: { ...this.state.formErrors, resetErrors },
+		} );
+	}
+
+	handleFieldChange = event => {
+		const formState = get( this.state, 'form', {} );
+		const { formErrors } = this.state;
+
+		formState[ event.target.name ] = event.target.value;
+		formErrors[ event.target.name ] = false;
+		this.setState( {
+			form: formState,
+			formErrors,
+		} );
+	};
+
+	handleSubmit = () => {
+		const { siteId, updateCredentials, translate } = this.props;
+
+		const payload = {
+			...this.state.form,
+			role: 'main',
+		};
+
+		let error = false;
+		const errors = this.state.formErrors;
+
+		if ( '' === payload.host ) {
+			errors.host = translate( 'Please enter a valid server address.' );
+			error = true;
+		}
+
+		if ( '' === payload.port ) {
+			errors.port = translate( 'Please enter a valid server port.' );
+			error = true;
+		}
+
+		if ( isNaN( payload.port ) ) {
+			errors.port = translate( 'Port number must be numeric.' );
+			error = true;
+		}
+
+		if ( '' === payload.user ) {
+			errors.user = translate( 'Please enter your server username.' );
+			error = true;
+		}
+
+		if ( '' === payload.pass ) {
+			errors.pass = translate( 'Please enter your server password.' );
+			error = true;
+		}
+
+		if ( error ) {
+			this.setState( { formErrors: errors } );
+			return;
+		}
+
+		updateCredentials( siteId, payload );
+	};
+
+	togglePublicKeyField = () =>
+		this.setState( { showPublicKeyField: ! this.state.showPublicKeyField } );
+
+	render() {
+		const { credentialsUpdating, onCancel, translate } = this.props;
+
+		const { showPublicKeyField, formErrors } = this.state;
+
+		return (
+			<FormFieldset className="credentials-form">
+				<table>
+					<tbody>
+						<tr>
+							<td colSpan="2" className="credentials-form__protocol-field">
+								<FormLabel>
+									<div>{ translate( 'Credential Type' ) }</div>
+									<FormSelect
+										name="protocol"
+										value={ get( this.state.form, 'protocol', 'ssh' ) }
+										onChange={ this.handleFieldChange }
+										disabled={ credentialsUpdating }
```

Comment:
this form doesn't really care about `credentialsUpdating` I don't think - I think it cares more about stuff like "form is submitting" or "form is disabled"

I mention this because as we start to step back and understand _what_ exactly we have, it becomes more general/reusable